### PR TITLE
Prepare GraphQL Crate for extraction

### PR DESCRIPTION
**Pull request by ControlCplusControlV**
This PR works to simplify the `query` interfaces by extracting the logic from all `*contexts` into traits, then refactors all queries to fetch this information via type objects rather than directly through the database, which will allow a clean removal of the graphql crate, where the traits will move to, while the implementations for `Database` remain within `fuel-core`.

There are some small pointer annoyances such as `borrowed_box`, but this is unavoidable because `Database` is passed into the graphql context as a `Box`, and so whenever its accessed the pointer to it by default is `&Box<T>`, and so any function which requires erasing `Box<DatabasePort>` (which is aliased to `Database` with the graphql crate) to something like `Box<dyn MessageProofData>` will inevitable create a borrowed box and so clippy will complain, but this can't really be avoided due to the graphql structure

### CODE REVIEW

**Comment by ControlCplusControlV** [comment]
Closes #985

File: crates/fuel-core/src/graphql_api/service.rs
Comment (xgreenx) [inline crates/fuel-core/src/graphql_api/service.rs:149]: We don't need to pass any `Box<dyn QueryData>`. The `Database` itself already should implement `QueryData` traits because of your blanket implementation.

And seems `QueryData` should be a super trait for `DatabasePort`.

File: crates/fuel-core/src/query.rs
Comment (xgreenx) [inline crates/fuel-core/src/query.rs:24]:
```suggestion
impl<D> QueryData for D 
where D:
     Send
     + Sync
     + BalanceQueryData
     + BlockQueryData
     + ChainQueryData
     + CoinQueryData
     + ContractQueryData
     + MessageQueryData
     + MessageProofData
     + TransactionQueryData
     {}
```

I guess it should be something like this

File: crates/fuel-core/src/query/balance/asset_query.rs
Comment (xgreenx) [inline crates/fuel-core/src/query/balance/asset_query.rs:102]: The output type should be `impl Iterator<Item = StorageResult<Resource>>`.

```
             .into_boxed()
             .collect::<Vec<_>>()
             .into_iter()
             .into_boxed()
```

Should definitely be re-worked=) I guess the previous implementation should work for you

File: crates/fuel-core/src/query/contract.rs
Comment (xgreenx) [inline crates/fuel-core/src/query/contract.rs:26]: Could you add `\n` after each method please?=)

File: crates/fuel-core/src/query/contract.rs
Comment (xgreenx) [inline crates/fuel-core/src/query/contract.rs:97]: You don't need to wrap it `into_boxed` because it is already wrapped=) 
image

Could you review all other places with `into_boxed` to avoid double wrapping, please?=)

File: crates/fuel-core/src/query/message.rs
Comment (xgreenx) [inline crates/fuel-core/src/query/message.rs:66]:
```suggestion
    ) -> BoxedIter<StorageResult<MessageId>>;
    
```

And for other methods too

File: crates/fuel-core/src/query/message.rs
Comment (xgreenx) [inline crates/fuel-core/src/query/message.rs:185]:
```suggestion
pub fn message_proof<'a, T: MessageProofData + ?Sized>(
    data: &T,
```

File: crates/fuel-core/src/query/message.rs
Comment (xgreenx) [inline crates/fuel-core/src/query/message.rs:168]: We don't need to duplicate functionality from `TransactionQueryData`, you simply can remove those function and use analogs from `TransactionQueryData`

File: crates/fuel-core/src/query/tx.rs
Comment (xgreenx) [inline crates/fuel-core/src/query/tx.rs:28]:
```suggestion
    fn transaction(&self, tx_id: &TxId) -> StorageResult<Transaction>;
    
```

And other methods

File: crates/fuel-core/src/resource_query.rs
Comment (xgreenx) [inline crates/fuel-core/src/resource_query.rs:98]:
```suggestion
```

File: crates/fuel-core/src/resource_query.rs
Comment (xgreenx) [inline crates/fuel-core/src/resource_query.rs:311]:
```suggestion
                        db,
```

File: crates/fuel-core/src/schema.rs
Comment (xgreenx) [inline crates/fuel-core/src/schema.rs:3]:
```suggestion
```

File: crates/fuel-core/src/schema/balance.rs
Comment (xgreenx) [inline crates/fuel-core/src/schema/balance.rs:55]:
```suggestion
        let data: &Database = ctx.data_unchecked();
```

File: crates/fuel-core/src/schema/balance.rs
Comment (xgreenx) [inline crates/fuel-core/src/schema/balance.rs:77]:
```suggestion
        let query: &Database = ctx.data_unchecked();
```

File: crates/fuel-core/src/schema/block.rs
Comment (xgreenx) [inline crates/fuel-core/src/schema/block.rs:92]:
```suggestion
        let query: &Database = ctx.data_unchecked();
```

And the same in other places=)

File: crates/fuel-core/src/schema/tx.rs
Comment (xgreenx) [inline crates/fuel-core/src/schema/tx.rs:276]: Hmm, the previous code with `get_tx_status` was better=) Could we return it back?

File: crates/fuel-core/src/service/sub_services.rs
Comment (xgreenx) [inline crates/fuel-core/src/service/sub_services.rs:177]: Yeah, we definitely don't need a separate `query_data` type=) Because it is `Database`

File: crates/storage/Cargo.toml
Comment (xgreenx) [inline crates/storage/Cargo.toml:24]: Only `fuel-core` should depend on the `rocksdb`. It is better to do this mapping in the `fuel-core`

**Review by xgreenx — DISMISSED** [review]

File: crates/fuel-core/src/query/balance/asset_query.rs
Comment (ControlCplusControlV) [reply crates/fuel-core/src/query/balance/asset_query.rs:102]: The output type of `impl Iterator<Item = StorageResult<Resource>>` can't be used as a return type in a trait definition, and referencing the iterator type directly is nearly impossible because it ends up being long enough that the compiler will dump its type name into a `.txt` file when compiling, is there another way around this?

File: crates/fuel-core/src/query/contract.rs
Comment (ControlCplusControlV) [reply crates/fuel-core/src/query/contract.rs:26]: Added :)

File: crates/fuel-core/src/query/balance/asset_query.rs
Comment (xgreenx) [reply crates/fuel-core/src/query/balance/asset_query.rs:102]: It is not a trait, it is a public method=)

Comment (ControlCplusControlV) [reply crates/fuel-core/src/query/balance/asset_query.rs]: Oh I had this confused from some of the other `impl Iterator`'s which I changed, thanks for fixing this!

File: crates/fuel-core/src/resource_query.rs
Comment (ControlCplusControlV) [reply crates/fuel-core/src/resource_query.rs:311]: fixed

File: crates/fuel-core/src/schema/balance.rs
Comment (ControlCplusControlV) [reply crates/fuel-core/src/schema/balance.rs]: Fixed

File: crates/fuel-core/src/query/message.rs
Comment (ControlCplusControlV) [reply crates/fuel-core/src/query/message.rs]: Resolved by moving directly to a GQL`Database` reference, sized issues among other things stopped me from doing just `&T`

File: crates/fuel-core/src/query/message.rs
Comment (ControlCplusControlV) [reply crates/fuel-core/src/query/message.rs:168]: It breaks a test which depends on `mockall` to gain access to these functions, but will refactor those tests

File: crates/storage/Cargo.toml
Comment (ControlCplusControlV) [reply crates/storage/Cargo.toml]: Made a wrapper type and fixed it

File: crates/fuel-core/src/graphql_api/service.rs
Comment (ControlCplusControlV) [reply crates/fuel-core/src/graphql_api/service.rs]: Removed QueryData

File: crates/fuel-core/src/schema/tx.rs
Comment (ControlCplusControlV) [reply crates/fuel-core/src/schema/tx.rs:276]: moved back

**Review by deekerno — COMMENTED** [review]
LGTM, but I'll defer to reviewers with more experience in this codebase.

File: crates/fuel-core/src/graphql_api.rs
Comment (ControlCplusControlV) [reply crates/fuel-core/src/graphql_api.rs]: Fixed, good catch

File: crates/fuel-core/src/schema/tx.rs
Comment (xgreenx) [inline crates/fuel-core/src/schema/tx.rs:273]:
```suggestion
        types::get_tx_status(id, self.db, self.txpool).await
```

File: crates/fuel-core/src/state/rocks_db.rs
Comment (xgreenx) [inline crates/fuel-core/src/state/rocks_db.rs:222]: If you use `DatabaseIterDirection` only here, it is better not to create this type and replace converting by some funciton=)

File: crates/fuel-core/src/schema/tx/types.rs
Comment (xgreenx) [inline crates/fuel-core/src/schema/tx/types.rs:405]: Why did you extract this functionality form the `get_tx_status` method?

Comment (ControlCplusControlV) [reply crates/fuel-core/src/schema/tx/types.rs:405]: There was a brief lifetime error where it wouldn't pass in that I fixed but forgot to change this back

File: crates/fuel-core/src/query/balance.rs
Comment (ControlCplusControlV) [reply crates/fuel-core/src/query/balance.rs:41]: This is unfortunately blocked by
```
        let amount = AssetQuery::new(
            &owner,
            &AssetSpendTarget::new(asset_id, u64::MAX, u64::MAX),
            None,
            self,
        )
 ```
 
 Just a couple lines below, as `self` cannot be passed alone because it is explicitly not sized, and even when trying to pass it as a reference I kept running into multiple sizing issues, and a Box pointer was the most simple way around this, if you do have a way that works locally feel free to commit it directly but I tried to get this to work both now and when writing it, and the sizing constraint blocks it from working

**Review by xgreenx — APPROVED** [review]

**Pull request merged**
